core/controllers/classifier_test.py

```diff
@@ -125,3 +126,78 @@ def test_error_on_invalid_message(self):
         self.payload['message']['job_id'] = 1
         self.post_json('/ml/trainedclassifierhandler', self.payload,
                        expect_errors=True, expected_status_int=400)
+
+    def test_error_on_existing_classifier(self):
+        # Create ClassifierDataModel before the controller is called.
+        classifier_services.create_classifier(self.job_id, self.classifier_data)
+        self.post_json('/ml/trainedclassifierhandler', self.payload,
+                       expect_errors=True, expected_status_int=500)
+
+
+class NextJobHandlerTest(test_utils.GenericTestBase):
+    """Test the handler for fetching next training job."""
+
+    def setUp(self):
+        super(NextJobHandlerTest, self).setUp()
+
+        self.exp_id = 'exp_id1'
+        self.title = 'Testing Classifier storing'
+        self.category = 'Test'
+        interaction_id = 'TextInput'
+        self.algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING[
+            interaction_id]['algorithm_id']
+        self.training_data = [
+            {
+                u'answer_group_index': 1,
+                u'answers': [u'a1', u'a2']
+            },
+            {
+                u'answer_group_index': 2,
+                u'answers': [u'a2', u'a3']
+            }
+        ]
+        self.job_id = classifier_services.create_classifier_training_job(
+            self.algorithm_id, interaction_id, self.exp_id,
+            1, 'Home', self.training_data,
+            feconf.TRAINING_JOB_STATUS_NEW)
+        self.expected_response = {
+            u'job_id' : unicode(self.job_id, 'utf-8'),
+            u'training_data' : self.training_data,
+            u'algorithm_id' : unicode(self.algorithm_id, 'utf-8')
+        }
+
+        self.payload = {}
+        self.payload['vm_id'] = feconf.DEFAULT_VM_ID
+        secret = feconf.DEFAULT_VM_SHARED_SECRET
+        self.payload['message'] = json.dumps({})
+        self.payload['signature'] = classifier.generate_signature(
+            secret, self.payload['message'])
+
+    def test_next_job_handler(self):
+        json_response = self.post_json('/ml/nextjobhandler',
+                                       self.payload, expect_errors=False,
+                                       expected_status_int=200)
+        self.assertEqual(json_response, self.expected_response)
+        classifier_services.mark_training_jobs_failed([self.job_id])
+        json_response = self.post_json('/ml/nextjobhandler',
+                                       self.payload, expect_errors=False,
+                                       expected_status_int=200)
+        self.assertEqual(json_response, {})
+
+    def test_error_on_prod_mode_and_default_vm_id(self):
+        # Turn off DEV_MODE.
+        with self.swap(feconf, 'DEV_MODE', False):
+            self.post_json('/ml/nextjobhandler', self.payload,
+                           expect_errors=True, expected_status_int=401)
+
+    def test_error_on_different_signatures(self):
```

Comment:
This and the next test looks the same. Perhaps, you wanted to have `test_error_on_modified_message` and modify the message or perhaps `test_error_on_invalid_signature` and modify the signature within the payload.